File: servidores/csharp/Servidor/Shared/RelogioProcesso.cs
using Google.Protobuf.WellKnownTypes;

namespace Shared;

public class RelogioProcesso
{
    private readonly object _sync = new();
    private long _logical = 0;
    private long _offsetNs = 0;

    public long BeforeSend()
    {
        lock (_sync)
        {
            _logical += 1;
            return _logical;
        }
    }

    public long OnReceive(long recebido)
    {
        lock (_sync)
        {
            _logical = Math.Max(_logical, recebido);
            return _logical;
        }
    }

    public long Valor()
    {
        lock (_sync)
        {
            return _logical;
        }
    }

    public Timestamp NowCorrigidoTimestamp()
    {
        lock (_sync)
        {
            return Mensageria.TimestampFromNs(Mensageria.AgoraNs() + _offsetNs);
        }
    }

    public long AtualizarOffset(Timestamp referencia, long envioNs, long recebimentoNs)
    {
        var pontoMedio = (envioNs + recebimentoNs) / 2L;
        var refNs = referencia.Seconds * 1_000_000_000L + referencia.Nanos;
        lock (_sync)
        {
            _offsetNs = refNs - pontoMedio;
            return _offsetNs;
        }
    }

    public double OffsetMillis()
    {
        lock (_sync)
        {
            return _offsetNs / 1_000_000.0;
        }
    }
}

File: shared/java/mensageria/RelogioProcesso.java
package shared.mensageria;

import com.google.protobuf.Timestamp;
import java.time.Instant;
import java.util.concurrent.atomic.AtomicLong;

public class RelogioProcesso {
  private final AtomicLong logical = new AtomicLong(0L);
  private final AtomicLong offsetNs = new AtomicLong(0L);

  public long beforeSend() {
    return logical.incrementAndGet();
  }

  public long onReceive(long recebido) {
    while (true) {
      long atual = logical.get();
      if (recebido <= atual) {
        return atual;
      }
      if (logical.compareAndSet(atual, recebido)) {
        return recebido;
      }
    }
  }

  public long valor() {
    return logical.get();
  }

  public Timestamp nowCorrigidoTimestamp() {
    long ns = Mensageria.agoraNs() + offsetNs.get();
    return Mensageria.timestampFromNs(ns);
  }

  public long atualizarOffset(Timestamp referencia, long envioNs, long recebimentoNs) {
    if (referencia == null) {
      return offsetNs.get();
    }
    long pontoMedio = (envioNs + recebimentoNs) / 2L;
    long refNs = referencia.getSeconds() * 1_000_000_000L + referencia.getNanos();
    long offset = refNs - pontoMedio;
    offsetNs.set(offset);
    return offset;
  }

  public double offsetMillis() {
    return offsetNs.get() / 1_000_000.0;
  }
}
